lib/mongo/error/operation_failure.rb

```diff
@@ -306,6 +316,28 @@ def unsupported_retryable_write?
         # either having string keys or providing indifferent access.
         code == 20 && server_message&.start_with?("Transaction numbers") || false
       end
+
+      private
+
+      # Retrieve the details from a document
+      #
+      # @return [ Hash | nil ] the details extracted from the document
+      def retrieve_details(document)
+        return nil unless document
+        if document['writeConcernError']
+          return document['writeConcernError']['errInfo']
+        elsif document['writeErrors'] && document['writeErrors'][0]
```

Comment:
Is there a way to do this that's more idiomatic to ruby?

Reply:
@Neilshweky Ruby's [safe navigation operator](https://mitrev.net/ruby/2015/11/13/the-operator-in-ruby/) might work for this:

```ruby
d1 = {}
d2 = {}
d2["writeErrors"] = [1,2,3]
d1
 => {} 
d2
=> {"writeErrors"=>[1, 2, 3]} 
d1["writeErrors"]&.first
=> nil
d2["writeErrors"]&.first
=> 1 
```

so maybe something along the lines of the following might work?

```ruby
elsif document['writeErrors']&.first
```

Reply:
You cound do `document['writeConcernError']&.[]('errInfo') || document['writeErrors']&.first&.[]('errInfo')` but whether this is more readable is open to interpretation.

I would assign intermediate values to variables to avoid repeated traversal for both runtime performance and (hopefully) ease of readability:

```

        if wce = document['writeConcernError']
          return wce['errInfo']
...
```